### PR TITLE
fix: allRowsSelected is false when there are no rows

### LINKED ISSUE #28927: [Bug]: Datagrid with selection and without items shows "all selected" in header
### Library

React Components / v9 (@fluentui/react-components)

### System Info

```shell
Browsers:
    Edge: Spartan (44.22621.2134.0), Chromium (115.0.1901.203)
```

### Are you reporting Accessibility issue?

None

### Reproduction

https://codesandbox.io/s/happy-rgb-cqv5kl?file=/example.tsx

## Actual Behavior

A selection list that does not contain any items should not indicate that all items are selected in the header. This may result in the header selection flashing briefly when items are in the process of loading: First the empty list appears as "all selected", then, once items are loaded and listed, the header selection reverts to the unselected state.

datagrid-no-item-selection

## Expected Behavior

If there are no items in the DataGrid, then the header checkbox should be in the unchecked state.

### Logs

_No response_

### Requested priority

Normal

### Products/sites affected

_No response_

### Are you willing to submit a PR to fix?

no

### Validations

- [X] Check that there isn't already an issue that reports the same bug to avoid creating a duplicate.
- [X] The provided reproduction is a minimal reproducible example of the bug.

### PR DESCRIPTION
**Pull request by ling1726**
Fixes #28927

### CODE REVIEW

File: packages/react-components/react-table/stories/DataGrid/Default.stories.tsx [hunk 1]
```diff
@@ -88,6 +88,8 @@ const items: Item[] = [
   },
 ];
 
+items = [];
```
Comment (layershifter) [inline packages/react-components/react-table/stories/DataGrid/Default.stories.tsx:91]: Restore this example?

**Review by layershifter — APPROVED** [review]

**Comment by fabricteam** [comment]
## [Perf Analysis (`@fluentui/react-components`)](https://github.com/microsoft/fluentui/wiki/Perf-Testing)
<p>No significant results to display.</p><details><summary>All results</summary><p>
  <table>
  <tr>
    <th>Scenario</th>
    <th>Render type</th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">Master Ticks</a>
    </th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">PR Ticks</a>
    </th>
    <th>Iterations</th>
    <th>Status</th>
  </tr><tr>
            <td>Avatar</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/30706/perf-test-react-components/Avatar-mount_base.html">656</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/30706/perf-test-react-components/Avatar-mount.html">643</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Button</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/30706/perf-test-react-components/Button-mount_base.html">305</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/30706/perf-test-react-components/Button-mount.html">295</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Field</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/30706/perf-test-react-components/Field-mount_base.html">1140</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/30706/perf-test-react-components/Field-mount.html">1158</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProvider</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/30706/perf-test-react-components/FluentProvider-mount_base.html">709</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/30706/perf-test-react-components/FluentProvider-mount.html">698</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/30706/perf-test-react-components/FluentProviderWithTheme-mount_base.html">83</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/30706/perf-test-react-components/FluentProviderWithTheme-mount.html">79</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/30706/perf-test-react-components/FluentProviderWithTheme-virtual-rerender_base.html">61</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/30706/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.html">62</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender-with-unmount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/30706/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount_base.html">77</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/30706/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount.html">67</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>MakeStyles</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/30706/perf-test-react-components/MakeStyles-mount_base.html">840</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/30706/perf-test-react-components/MakeStyles-mount.html">897</a></td>
            <td>50000</td>
            <td></td>
           </tr>
<tr>
            <td>Persona</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/30706/perf-test-react-components/Persona-mount_base.html">1746</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/30706/perf-test-react-components/Persona-mount.html">1731</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>SpinButton</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/30706/perf-test-react-components/SpinButton-mount_base.html">1413</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/30706/perf-test-react-components/SpinButton-mount.html">1348</a></td>
            <td>5000</td>
            <td></td>
           </tr></table></p></details>

<div id="perfComment9425"/>

**Comment by fabricteam** [comment]
## 📊 Bundle size report

| Package & Exports | Baseline (minified/GZIP) | PR    | Change     |
| :---------------- | -----------------------: | ----: | ---------: |
| <samp>react-table</samp> <br /> <abbr title='bundle-size/DataGrid.fixture.js'>DataGrid</abbr> | `165.168 kB`<br />`45.902 kB` | `165.191 kB`<br />`45.909 kB` | `23 B` <br />`7 B` |
| <samp>react-table</samp> <br /> <abbr title='bundle-size/TableAsDataGrid.fixture.js'>Table as DataGrid</abbr> | `136.55 kB`<br />`36.811 kB` | `136.573 kB`<br />`36.817 kB` | `23 B` <br />`6 B` |
| <samp>react-table</samp> <br /> <abbr title='bundle-size/TableSelectionOnly.fixture.js'>Table (Selection only)</abbr> | `76.311 kB`<br />`20.546 kB` | `76.334 kB`<br />`20.552 kB` | `23 B` <br />`6 B` |

<details>
<summary>Unchanged fixtures</summary>

| Package & Exports | Size (minified/GZIP) |
| ----------------- | -------------------: |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/ButtonProviderAndTheme.fixture.js'>react-components: Button, FluentProvider & webLightTheme</abbr> | `71.098 kB`<br />`20.514 kB` |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/MultipleComponents.fixture.js'>react-components: Accordion, Button, FluentProvider, Image, Menu, Popover</abbr> | `217.207 kB`<br />`61.943 kB` |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/ProviderAndTheme.fixture.js'>react-components: FluentProvider & webLightTheme</abbr> | `43.585 kB`<br />`14.351 kB` |
| <samp>react-portal-compat</samp> <br /> <abbr title='bundle-size/PortalCompat.fixture.js'>PortalCompatProvider</abbr> | `7.944 kB`<br />`2.588 kB` |
| <samp>react-table</samp> <br /> <abbr title='bundle-size/PrimitivesOnly.fixture.js'>Table (Primitives only)</abbr> | `45.324 kB`<br />`14.116 kB` |
| <samp>react-table</samp> <br /> <abbr title='bundle-size/TableSortOnly.fixture.js'>Table (Sort only)</abbr> | `74.977 kB`<br />`20.154 kB` |
</details>
<sub>🤖 This report was generated against <a href='https://github.com/microsoft/fluentui/commit/1fa9745b9155c4fec5fe435acc29eee0b40eb695'>1fa9745b9155c4fec5fe435acc29eee0b40eb695</a></sub>
<div id="bundleSizeComment9423"/>

**Comment by fabricteam** [comment]
## 🕵 fluentuiv9 No visual regressions between this PR and main
<div id="vrtCommentfluent-ui_VRT_Pipeline_v9"/>

**Pull request merged**